### PR TITLE
Add a fast algorithm for AC of AllDiff.

### PR DESCRIPTION
**Pull request by amtfjlsj**
Add Zhang algorithm in the paper of IJCAI-18 "A Fast Algorithm for Generalized Arc Consistency of the Alldifferent Constraint".
 
@chocoteam/core-developer

### CODE REVIEW

**Comment by jgFages** [comment]
Hi, thank you for the contribution. Could you please bring some tests showing this implementation is indeed faster than the existing one? Thank you in advance

**Comment by amtfjlsj** [comment]
> 
> 
> Hi, thank you for the contribution. Could you please bring some tests showing this implementation is indeed faster than the existing one? Thank you in advance

No problem. I have tested some instances of which all constraints are AllDiff from XCSP3. The experiment uses the default heuristic of Choco and the timeout is 180s. The result is shown below.

File: src/main/java/org/chocosolver/util/graphOperations/connectivity/StrongConnectivityFinder.java [hunk 1]
```diff
@@ -69,6 +69,17 @@ public void findAllSCC() {
 		findAllSCCOf(restriction);
 	}
 
+	// exception is a set of nodes that do not need to be found SCC
+	public void findAllSCC(BitSet exception) {
+		ISet nodes = graph.getNodes();
+		for (int i = 0; i < n; i++) {
```
[the remark this answers is not in the document]
Comment (amtfjlsj) [reply src/main/java/org/chocosolver/util/graphOperations/connectivity/StrongConnectivityFinder.java]: This is a good suggestion. After I test it, the right code is:
`for (int i = exception.nextClearBit(0); i >= 0 && i < n; i = exception.nextClearBit(i + 1)) {`

File: src/main/java/org/chocosolver/util/graphOperations/connectivity/StrongConnectivityFinder.java [hunk 2]
```diff
@@ -88,9 +99,8 @@ public void findAllSCCOf(BitSet restriction) {
 	}
 
 	private void findSingletons(BitSet restriction) {
-		ISet nodes = graph.getNodes();
 		for (int i = restriction.nextSetBit(0); i >= 0; i = restriction.nextSetBit(i + 1)) {
-			if (nodes.contains(i) && graph.getPredOf(i).size() * graph.getSuccOf(i).size() == 0) {
+			if (graph.getPredOf(i).size() * graph.getSuccOf(i).size() == 0) {
```
[the remark this answers is not in the document]
Comment (amtfjlsj) [reply src/main/java/org/chocosolver/util/graphOperations/connectivity/StrongConnectivityFinder.java]: OK

File: src/main/java/org/chocosolver/solver/constraints/nary/alldifferent/algo/AlgoAllDiffACFast.java [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    //***********************************************************************************
+    // CONSTRUCTORS
+    //***********************************************************************************
+
+    public AlgoAllDiffACFast(IntVar[] variables, ICause cause) {
+        this.vars = variables;
+        aCause = cause;
+        n = vars.length;
+        matching = new int[n];
+        for (int i = 0; i < n; i++) {
```
[the remark this answers is not in the document]
Comment (amtfjlsj) [reply src/main/java/org/chocosolver/solver/constraints/nary/alldifferent/algo/AlgoAllDiffACFast.java]: OK

**Comment by cprudhom** [comment]
I integrated your code in `modules` branch.
It should be merged into master in few days.
Once again, thank you for your contribution.

Best,
CP

**Pull request closed without merging**